### PR TITLE
Add ability to get whole diff

### LINKED ISSUE #814: Blacklisting file paths for lines_of_code
Is there a way to blacklist a set of file paths to efficiently using the `lines_of_code` API?

I would like to warn about excessive changes of lines of code change on a repo, but excludes changes done for tests, third-party dependency, and more. Does this make sense to others?

### PR DESCRIPTION
**Pull request by sleekybadger**
### Summary

Exposes `diff` object to Dangerfile dsl. Resolves #814 .

### Example 

```
p git.diff 
=> 
#<Git::Diff:0x007fcb8f9a1ae8
 @base=
  #<Git::Base:0x007fcb8fa27080
   @index=
    #<Git::Index:0x007fcb8fa26c70
     @path="/Users/sleekybadger/code/danger_test/.git/index">,
   @lib=
    #<Git::Lib:0x007fcb8f99dc18
     @git_dir="/Users/sleekybadger/code/danger_test/.git",
     @git_index_file="/Users/sleekybadger/code/danger_test/.git/index",
     @git_system_env_variables=
      {"GIT_DIR"=>nil,
       "GIT_WORK_TREE"=>nil,
       "GIT_INDEX_FILE"=>nil,
       "GIT_SSH"=>nil},
     @git_work_dir="/Users/sleekybadger/code/danger_test",
     @logger=nil,
     @path=nil>,
   @logger=nil,
   @repository=
    #<Git::Repository:0x007fcb8fa26d88
     @path="/Users/sleekybadger/code/danger_test/.git">,
   @working_directory=
    #<Git::WorkingDirectory:0x007fcb8fa26e28
     @path="/Users/sleekybadger/code/danger_test">>,
 @from="c837138f9a3c06294686c544af326a00ed663ed1",
 @full_diff=
  "diff --git a/Dangerfile b/Dangerfile\n" +
  "index e6aed66..f27e2f8 100644\n" +
  "--- a/Dangerfile\n" +
  "+++ b/Dangerfile\n" +
  "@@ -1 +1,3 @@\n" +
  "-message(\"Hi\")\n" +
  "+require \"pp\"\n" +
  "+\n" +
  "+pp git.diff\n" +
  "diff --git a/Gemfile b/Gemfile\n" +
  "index a432dd0..ffddea0 100644\n" +
  "--- a/Gemfile\n" +
  "+++ b/Gemfile\n" +
  "@@ -1,3 +1,5 @@\n" +
  " source \"https://rubygems.org\"\n" +
  " \n" +
  "-gem 'danger'\n" +
  "+gem 'danger', github: 'sleekybadger/danger',\n" +
  "+              branch: 'feature/git-diff',\n" +
  "+              ref: 'b53a9f18be4e26a88a2b0904356fe13e124966d4'\n" +
  "diff --git a/Gemfile.lock b/Gemfile.lock\n" +
  "index 3ecb6b8..14d3d88 100644\n" +
  "--- a/Gemfile.lock\n" +
  "+++ b/Gemfile.lock\n" +
  "@@ -1,3 +1,21 @@\n" +
  "+GIT\n" +
  "+  remote: git://github.com/sleekybadger/danger.git\n" +
  "+  revision: b53a9f18be4e26a88a2b0904356fe13e124966d4\n" +
  "+  ref: b53a9f18be4e26a88a2b0904356fe13e124966d4\n" +
  "+  branch: feature/git-diff\n" +
  "+  specs:\n" +
  "+    danger (5.2.2)\n" +
  "+      claide (~> 1.0)\n" +
  "+      claide-plugins (>= 0.9.2)\n" +
  "+      colored2 (~> 3.1)\n" +
  "+      cork (~> 0.1)\n" +
  "+      faraday (~> 0.9)\n" +
  "+      faraday-http-cache (~> 1.0)\n" +
  "+      git (~> 1)\n" +
  "+      kramdown (~> 1.5)\n" +
  "+      octokit (~> 4.2)\n" +
  "+      terminal-table (~> 1)\n" +
  "+\n" +
  " GEM\n" +
  "   remote: https://rubygems.org/\n" +
  "   specs:\n" +
  "@@ -11,17 +29,6 @@ GEM\n" +
  "     colored2 (3.1.2)\n" +
  "     cork (0.3.0)\n" +
  "       colored2 (~> 3.1)\n" +
  "-    danger (5.1.1)\n" +
  "-      claide (~> 1.0)\n" +
  "-      claide-plugins (>= 0.9.2)\n" +
  "-      colored2 (~> 3.1)\n" +
  "-      cork (~> 0.1)\n" +
  "-      faraday (~> 0.9)\n" +
  "-      faraday-http-cache (~> 1.0)\n" +
  "-      git (~> 1)\n" +
  "-      kramdown (~> 1.5)\n" +
  "-      octokit (~> 4.2)\n" +
  "-      terminal-table (~> 1)\n" +
  "     faraday (0.12.1)\n" +
  "       multipart-post (>= 1.2, < 3)\n" +
  "     faraday-http-cache (1.3.1)\n" +
  "@@ -37,15 +44,15 @@ GEM\n" +
  "     sawyer (0.8.1)\n" +
  "       addressable (>= 2.3.5, < 2.6)\n" +
  "       faraday (~> 0.8, < 1.0)\n" +
  "-    terminal-table (1.7.3)\n" +
  "-      unicode-display_width (~> 1.1.1)\n" +
  "-    unicode-display_width (1.1.3)\n" +
  "+    terminal-table (1.8.0)\n" +
  "+      unicode-display_width (~> 1.1, >= 1.1.1)\n" +
  "+    unicode-display_width (1.2.1)\n" +
  " \n" +
  " PLATFORMS\n" +
  "   ruby\n" +
  " \n" +
  " DEPENDENCIES\n" +
  "-  danger\n" +
  "+  danger!\n" +
  " \n" +
  " BUNDLED WITH\n" +
  "    1.14.6",
 @full_diff_files=
  [["Dangerfile",
    #<Git::Diff::DiffFile:0x007fcb8f070e38
     @base=
      #<Git::Base:0x007fcb8fa27080
       @index=
        #<Git::Index:0x007fcb8fa26c70
         @path="/Users/sleekybadger/code/danger_test/.git/index">,
       @lib=
        #<Git::Lib:0x007fcb8f99dc18
         @git_dir="/Users/sleekybadger/code/danger_test/.git",
         @git_index_file="/Users/sleekybadger/code/danger_test/.git/index",
         @git_system_env_variables=
          {"GIT_DIR"=>nil,
           "GIT_WORK_TREE"=>nil,
           "GIT_INDEX_FILE"=>nil,
           "GIT_SSH"=>nil},
         @git_work_dir="/Users/sleekybadger/code/danger_test",
         @logger=nil,
         @path=nil>,
       @logger=nil,
       @repository=
        #<Git::Repository:0x007fcb8fa26d88
         @path="/Users/sleekybadger/code/danger_test/.git">,
       @working_directory=
        #<Git::WorkingDirectory:0x007fcb8fa26e28
         @path="/Users/sleekybadger/code/danger_test">>,
     @binary=nil,
     @dst="f27e2f8",
     @mode="100644",
     @patch=
      "diff --git a/Dangerfile b/Dangerfile\n" +
      "index e6aed66..f27e2f8 100644\n" +
      "--- a/Dangerfile\n" +
      "+++ b/Dangerfile\n" +
      "@@ -1 +1,3 @@\n" +
      "-message(\"Hi\")\n" +
      "+require \"pp\"\n" +
      "+\n" +
      "+pp git.diff",
     @path="Dangerfile",
     @src="e6aed66",
     @type="modified">],
   ["Gemfile",
    #<Git::Diff::DiffFile:0x007fcb8f070ca8
     @base=
      #<Git::Base:0x007fcb8fa27080
       @index=
        #<Git::Index:0x007fcb8fa26c70
         @path="/Users/sleekybadger/code/danger_test/.git/index">,
       @lib=
        #<Git::Lib:0x007fcb8f99dc18
         @git_dir="/Users/sleekybadger/code/danger_test/.git",
         @git_index_file="/Users/sleekybadger/code/danger_test/.git/index",
         @git_system_env_variables=
          {"GIT_DIR"=>nil,
           "GIT_WORK_TREE"=>nil,
           "GIT_INDEX_FILE"=>nil,
           "GIT_SSH"=>nil},
         @git_work_dir="/Users/sleekybadger/code/danger_test",
         @logger=nil,
         @path=nil>,
       @logger=nil,
       @repository=
        #<Git::Repository:0x007fcb8fa26d88
         @path="/Users/sleekybadger/code/danger_test/.git">,
       @working_directory=
        #<Git::WorkingDirectory:0x007fcb8fa26e28
         @path="/Users/sleekybadger/code/danger_test">>,
     @binary=nil,
     @dst="ffddea0",
     @mode="100644",
     @patch=
      "diff --git a/Gemfile b/Gemfile\n" +
      "index a432dd0..ffddea0 100644\n" +
      "--- a/Gemfile\n" +
      "+++ b/Gemfile\n" +
      "@@ -1,3 +1,5 @@\n" +
      " source \"https://rubygems.org\"\n" +
      " \n" +
      "-gem 'danger'\n" +
      "+gem 'danger', github: 'sleekybadger/danger',\n" +
      "+              branch: 'feature/git-diff',\n" +
      "+              ref: 'b53a9f18be4e26a88a2b0904356fe13e124966d4'",
     @path="Gemfile",
     @src="a432dd0",
     @type="modified">],
   ["Gemfile.lock",
    #<Git::Diff::DiffFile:0x007fcb8f070bb8
     @base=
      #<Git::Base:0x007fcb8fa27080
       @index=
        #<Git::Index:0x007fcb8fa26c70
         @path="/Users/sleekybadger/code/danger_test/.git/index">,
       @lib=
        #<Git::Lib:0x007fcb8f99dc18
         @git_dir="/Users/sleekybadger/code/danger_test/.git",
         @git_index_file="/Users/sleekybadger/code/danger_test/.git/index",
         @git_system_env_variables=
          {"GIT_DIR"=>nil,
           "GIT_WORK_TREE"=>nil,
           "GIT_INDEX_FILE"=>nil,
           "GIT_SSH"=>nil},
         @git_work_dir="/Users/sleekybadger/code/danger_test",
         @logger=nil,
         @path=nil>,
       @logger=nil,
       @repository=
        #<Git::Repository:0x007fcb8fa26d88
         @path="/Users/sleekybadger/code/danger_test/.git">,
       @working_directory=
        #<Git::WorkingDirectory:0x007fcb8fa26e28
         @path="/Users/sleekybadger/code/danger_test">>,
     @binary=nil,
     @dst="14d3d88",
     @mode="100644",
     @patch=
      "diff --git a/Gemfile.lock b/Gemfile.lock\n" +
      "index 3ecb6b8..14d3d88 100644\n" +
      "--- a/Gemfile.lock\n" +
      "+++ b/Gemfile.lock\n" +
      "@@ -1,3 +1,21 @@\n" +
      "+GIT\n" +
      "+  remote: git://github.com/sleekybadger/danger.git\n" +
      "+  revision: b53a9f18be4e26a88a2b0904356fe13e124966d4\n" +
      "+  ref: b53a9f18be4e26a88a2b0904356fe13e124966d4\n" +
      "+  branch: feature/git-diff\n" +
      "+  specs:\n" +
      "+    danger (5.2.2)\n" +
      "+      claide (~> 1.0)\n" +
      "+      claide-plugins (>= 0.9.2)\n" +
      "+      colored2 (~> 3.1)\n" +
      "+      cork (~> 0.1)\n" +
      "+      faraday (~> 0.9)\n" +
      "+      faraday-http-cache (~> 1.0)\n" +
      "+      git (~> 1)\n" +
      "+      kramdown (~> 1.5)\n" +
      "+      octokit (~> 4.2)\n" +
      "+      terminal-table (~> 1)\n" +
      "+\n" +
      " GEM\n" +
      "   remote: https://rubygems.org/\n" +
      "   specs:\n" +
      "@@ -11,17 +29,6 @@ GEM\n" +
      "     colored2 (3.1.2)\n" +
      "     cork (0.3.0)\n" +
      "       colored2 (~> 3.1)\n" +
      "-    danger (5.1.1)\n" +
      "-      claide (~> 1.0)\n" +
      "-      claide-plugins (>= 0.9.2)\n" +
      "-      colored2 (~> 3.1)\n" +
      "-      cork (~> 0.1)\n" +
      "-      faraday (~> 0.9)\n" +
      "-      faraday-http-cache (~> 1.0)\n" +
      "-      git (~> 1)\n" +
      "-      kramdown (~> 1.5)\n" +
      "-      octokit (~> 4.2)\n" +
      "-      terminal-table (~> 1)\n" +
      "     faraday (0.12.1)\n" +
      "       multipart-post (>= 1.2, < 3)\n" +
      "     faraday-http-cache (1.3.1)\n" +
      "@@ -37,15 +44,15 @@ GEM\n" +
      "     sawyer (0.8.1)\n" +
      "       addressable (>= 2.3.5, < 2.6)\n" +
      "       faraday (~> 0.8, < 1.0)\n" +
      "-    terminal-table (1.7.3)\n" +
      "-      unicode-display_width (~> 1.1.1)\n" +
      "-    unicode-display_width (1.1.3)\n" +
      "+    terminal-table (1.8.0)\n" +
      "+      unicode-display_width (~> 1.1, >= 1.1.1)\n" +
      "+    unicode-display_width (1.2.1)\n" +
      " \n" +
      " PLATFORMS\n" +
      "   ruby\n" +
      " \n" +
      " DEPENDENCIES\n" +
      "-  danger\n" +
      "+  danger!\n" +
      " \n" +
      " BUNDLED WITH\n" +
      "    1.14.6",
     @path="Gemfile.lock",
     @src="3ecb6b8",
     @type="modified">]],
 @path=nil,
 @stats=
  {:total=>{:insertions=>28, :deletions=>17, :lines=>45, :files=>3},
   :files=>
    {"Dangerfile"=>{:insertions=>3, :deletions=>1},
     "Gemfile"=>{:insertions=>3, :deletions=>1},
     "Gemfile.lock"=>{:insertions=>22, :deletions=>15}}},
 @to="danger_head"
>
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 * Add your own contributions to the next release on the line below this, please include your name too. Please don't set a new version if you are the first to make the section for `master`.
 
 * Improve remotes parsing for init command - [@sleekybadger](https://github.com/sleekybadger)
+* Add ability to get whole diff - [@sleekybadger](https://github.com/sleekybadger)
```
Comment (orta) [inline CHANGELOG.md]: Any chance we can expand this a bit more: `Adds ability to get the whole diff object at `git.diff` - this lets you do more fine grained checks on the raw changes.`

Then people will know how to use it

Comment (sleekybadger) [inline CHANGELOG.md]: Sure 😉

**Comment by sleekybadger** [comment]
@orta updated

**Comment by orta** [comment]
🚢

**Pull request merged**